PointerArgumentPromotionPass.cpp

```diff
@@ -0,0 +1,228 @@
+    }
+    ArrayRef<Type *> NewArgTypes(NewArgs);
+
+    FunctionType *NFTy = FunctionType::get(RetTy, NewArgTypes, false);
+
+    Function *NF = Function::Create(NFTy, F->getLinkage());
+
+    NF->copyAttributesFrom(F);
+
+    if (F->getAlignment())
+      NF->setAlignment(Align(F->getAlignment()));
+    NF->setCallingConv(F->getCallingConv());
+    NF->setLinkage(F->getLinkage());
+    NF->setDSOLocal(F->isDSOLocal());
+    NF->setSubprogram(F->getSubprogram());
+    NF->setUnnamedAddr(F->getUnnamedAddr());
+    NF->setVisibility(F->getVisibility());
+    if (F->hasPersonalityFn())
+      NF->setPersonalityFn(F->getPersonalityFn());
+    if (F->hasComdat())
+      NF->setComdat(F->getComdat());
+    if (F->hasSection())
+      NF->setSection(F->getSection());
+
+    F->getParent()->getFunctionList().insert(F->getIterator(), NF);
+    NF->takeName(F);
+
+    // Since we have now created the new function, splice the body of the old
+    // function right into the new function, leaving the old rotting hulk of the
+    // function empty.
+    NF->getBasicBlockList().splice(NF->begin(), F->getBasicBlockList());
+
+    std::vector<Argument *> NewArgsList;
+    for (Argument &arg : NF->args()) {
+      NewArgsList.push_back(&arg);
+    }
+
+    // Loop over the argument list, transferring uses of the old arguments over to
+    // the new arguments, also transferring over the names as well. 
+
+    for (unsigned i = 0; i < NewArgsList.size(); i++) {
+      if (ArgPtrTy.find(OldArgsList[i])!=ArgPtrTy.end()) {
+        for (auto It = OldArgsList[i]->user_begin(); It!=OldArgsList[i]->user_end();) {
+          User *U = *It;
+          It++;
+          if (auto *I2P = dyn_cast<IntToPtrInst>(U)) {
+            IRBuilder<> Builder(I2P);
+            auto *CastPtr = Builder.CreatePointerCast(NewArgsList[i], I2P->getType());
+            I2P->replaceAllUsesWith(CastPtr);
+            I2P->eraseFromParent();
+          } else if (auto *RI = dyn_cast<ReturnInst>(U)) {
+            IRBuilder<> Builder(RI);
+            if (RetTy->isPointerTy())
+              RI->setOperand(0,Builder.CreatePointerCast(NewArgsList[i], RetTy));
+            else
+              RI->setOperand(0,Builder.CreatePtrToInt(NewArgsList[i], RetTy));
+          } else {
+            assert(false && "Unexpected instruction");
+          }
+        }
+      } else
+        OldArgsList[i]->replaceAllUsesWith(NewArgsList[i]);
+
+      NewArgsList[i]->takeName(OldArgsList[i]);
+    }
+
+    // Clone metadatas from the old function, including debug info descriptor.
+    SmallVector<std::pair<unsigned, MDNode *>, 1> MDs;
+    F->getAllMetadata(MDs);
+    for (auto MD : MDs)
+      NF->addMetadata(MD.first, *MD.second);
+
+    for (auto It = F->user_begin(); It!=F->user_end();) {
+      User *U = *It;
+      It++;
+
+      if (auto *CB = dyn_cast<CallBase>(U)) {
+        IRBuilder<> Builder(CB);
+
+        std::vector<Value*> Args;
+        for (unsigned i = 0; i<CB->getNumArgOperands(); i++) {
```

Comment:
`CallBase::getNumArgOperands` was removed in commit https://github.com/llvm/llvm-project/commit/3e1c787b3160bed4146d3b2b5f922aeed3caafd7.

I think this can be replaced with `CallBase ::arg_size`.
```suggestion
        for (unsigned i = 0; i<CB->arg_size(); i++) {
```